### PR TITLE
Bind coredns containers to  linux nodes to avoid Windows scheduling

### LINKED ISSUE #69773: Official coredns should be bound to nodeSelector: beta.kubernetes.io/os=linux
**Is this a BUG REPORT or FEATURE REQUEST?**: 
/kind bug

**What happened**:

In a windows cluster, containers can get scheduled on Windows nodes where they will never work.

Warning Failed 54m (x3 over 54m) kubelet, 35934k8s9000 Failed to pull image "coredns/coredns:1.2.2": rpc error: code = Unknown desc = no matching manifest for windows/amd64 in the manifest list entries

**What you expected to happen**:

CoreDNS starts and runs on Linux master nodes.

**How to reproduce it (as minimally and precisely as possible)**:

Build a Windows Cluster.

Note:  I first encountered this issue when deploying with ACS-Engine:

https://github.com/Azure/acs-engine/issues/4011

but it occurs to me that it should be mirrored here, so that the relevant node selectors can be included in the canonical add-on directory.

/sig Windows

### PR DESCRIPTION
**Pull request by MarcPow**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #69773

**Special notes for your reviewer**:

This prevents scheduling attempts that are sure to fail on Windows agents.  Given the current set of Windows issues, scheduling attempts may sometimes hang and prevent DNS from coming up cleanly in a cluster.

**Release note**:

```release-note
CoreDNS is only officially supported on Linux at this time.  As such, when kubeadm is used to deploy this component into your kubernetes cluster, it will be restricted (using nodeSelectors) to run only on nodes with that operating system. This ensures that in clusters which include Windows nodes, the scheduler will not ever attempt to place CoreDNS pods on these machines, reducing setup latency and enhancing initial cluster stability.
```

### CODE REVIEW

**Comment by MarcPow** [comment]
/sig Windows 

This is a micro-improvement to Windows clusters at startup time.

File: cluster/addons/dns/coredns/coredns.yaml.base
Comment (MrHohn) [inline cluster/addons/dns/coredns/coredns.yaml.base]: Is this label guaranteed to be present on node? When was this introduced?

**Comment by MrHohn** [comment]
@chrisohaver @rajansandeep do you mind taking a look? Thanks.

**Comment by chrisohaver** [comment]
Is CoreDNS the only core k8s component that cant run on windows?

**Comment by MarcPow** [comment]
> Is CoreDNS the only core k8s component that cant run on windows?

I don't have sufficient expertise (yet) to comment on that.   I just happened to get bitten by this one when setting a cluster up on a slightly older environment in Azure (where a combination of [seperately filed] bugs in the Docker EE + Azure CNI stacks led to the container to get 'stuck' on a particular node), leaving the cluster without DNS.

**Review by chrisohaver — APPROVED** [review]
Makes sense to me.  CoreDNS can't run on windows nodes currently.

The change should be proposed in other places too... 
* kubernetes/kubeadm (has it's own copy of the coredns manifests)
* coredns/deployment (intended to be the authoritative k8s deployment default configuration)

File: cluster/addons/dns/coredns/coredns.yaml.in
Comment (rajansandeep) [inline cluster/addons/dns/coredns/coredns.yaml.in]: @MarcPow please add these lines [here](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/addons/dns/manifests.go#L240) as well.

**Comment by rajansandeep** [comment]
> The change should be proposed in other places too...
@chrisohaver  I'll update them in other install tools and in `coredns/deployment`
/ok-to-test

File: cluster/addons/dns/coredns/coredns.yaml.in
Comment (MarcPow) [reply cluster/addons/dns/coredns/coredns.yaml.in]: Done.

File: cluster/addons/dns/coredns/coredns.yaml.base
Comment (MarcPow) [reply cluster/addons/dns/coredns/coredns.yaml.base]: Just to sanity check myself out of pure paranoia, the value of this default label comes from goruntime.GOOS.

Gere are the permissible values as I understand them - 
[https://github.com/golang/go/blob/master/src/go/build/syslist.go]
const goosList = "aix android darwin dragonfly freebsd js linux nacl netbsd openbsd plan9 solaris windows zos "

I believe "linux" is the only value pertinent to coredns, given that the official Dockerfile is bound to debian.
[https://github.com/coredns/coredns/blob/master/Dockerfile]

**Comment by MarcPow** [comment]
Test failures related to pre-existing flaky test: https://github.com/kubernetes/kubernetes/issues/66542

**Comment by rajansandeep** [comment]
@MarcPow can you cleanup and squash the commits?

**Comment by fabriziopandini** [comment]
@MarcPow did you considered `beta.kubernetes.io/arch=amd64` ?
Are there plans for multiarch / fat images support similar to what is happening for other k8s components?

**Comment by chrisohaver** [comment]
@MarcPow,

> Given the current set of Windows issues, scheduling attempts may sometimes hang and prevent DNS from coming up cleanly in a cluster.

Can you point to related issues or describe what these issues are, and in what way they affect scheduling or DNS coming up cleanly.  Why is this "may sometimes"?

**Comment by chrisohaver** [comment]
> ... did you considered beta.kubernetes.io/arch=amd64

the coredns image manifest has multi arch support for amd64, arm, arm64, ppc64le, and s390x.

**Comment by chrisohaver** [comment]
Does the kube-dns manifest need the same node selector?

**Comment by chrisohaver** [comment]
I was thinking that in a windows/linux hybrid cluster, windows nodes exist exclusively to run windows-only Pods, which means that it might make more sense to put a no-schedule taint on all windows nodes, and then select windows nodes only (tolerating the no-schedule taint), rather than to expressly forbid each linux-only pod from running on windows nodes.

**Comment by chrisohaver** [comment]
cc @daschott

**Comment by MarcPow** [comment]
> I was thinking that in a windows/linux hybrid cluster, windows nodes exist exclusively to run windows-only Pods, which means that it might make more sense to put a no-schedule taint on all windows nodes, and tolerate the no-schedule taint for windows-only Pods, rather than to expressly forbid each linux-only pod from running on windows nodes.

That feels strange to me, on balance.    If you have Pods that can only run on one OS, they should get scheduled only on nodes running that OS.     You could even argue that the scheduler should do that dynamically, by inspecting the image metadata, rather than doing this through explicit selectors...

**Comment by MarcPow** [comment]
> @MarcPow can you cleanup and squash the commits?

Done.

**Comment by MarcPow** [comment]
/ping 

Kubeadm approvers - any thoughts?

**Comment by MarcPow** [comment]
@fabriziopandini @luxas - This isn't able to be merged without your signoff on the kubeadm change.

**Comment by neolit123** [comment]
@MarcPow

please add a release note explaining the change.
```
kubeadm: bind CoreDNS containers to Linux nodes to avoid Windows scheduling.
```

@timothysc we can have this enabled for now i guess... CoreDNS does not work on Windows yet, as confirmed by @chrisohaver

**Review by timothysc — APPROVED** [review]
/approve

File: cmd/kubeadm/app/phases/addons/dns/manifests.go
Comment (MarcPow) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go]: That's a fair point - I'd love to see the scheduler to inspect the relevant images and dynamically impose these constraints.  But in the absence of such a feature, this seems like a reasonable place to start.

**Comment by chrisohaver** [comment]
> CoreDNS does not work on Windows yet

To clarify, CoreDNS itself works fine on Windows.  But we don't compile a windows executable into the CoreDNS manifest (AFAIK neither does kube-dns, so we probably want to change that as well).  

As I understand it, there are some hurdles in generally getting K8s cluster DNS on Windows (not coredns specific). A significant one is that Windows systems don't have an `/etc/resolv.conf`, and _kubelet_ is geared toward assuming all pods do.  So there needs to be some way of informing windows Pods of the DNS policy.

File: cmd/kubeadm/app/phases/addons/dns/manifests.go
Comment (neolit123) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go]: probable best to have an issue in the k/kubeadm repo about this and add a TODO comment in the code if this change is to make it in.

**Comment by neolit123** [comment]
> To clarify, CoreDNS itself works fine on Windows. 

there is a new WG in k8s that is currently discussed on how multi-platform images are to be build and tagged.

File: cmd/kubeadm/app/phases/addons/dns/manifests.go
Comment (timothysc) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go]: @brendanburns - ping!  What's the policy you are using for denoting windows nodes? and is there a KEP somewhere? 

/cc @jbeda

**Comment by chrisohaver** [comment]
IMO, If it makes sense to add this node selector for the CoreDNS deployment manifests, it should also be added to the kube-dns deployment manifests.  AFAIK, both have the same issue.

**Comment by timothysc** [comment]
I'm taking off approval, we don't have well defined policies on heterogeneous clustering that I'm aware of.

**Comment by daschott** [comment]
There are no core-DNS pods based off of windowsservercore container images, so I do think we need a way to schedule to the right OS. The current thoughts are captured here:
https://docs.google.com/document/d/1XLs8Mbz1-xOIiDW9XSSuhx9fshpxJM1NDD1a0oVbzfc/edit

and I'm sure it will be discussed again in future SIG-Windows meetings.

**Comment by neolit123** [comment]
what is preventing the workaround of updating the deployment with amended node selector toleration?
in kubeadm we had reports that CNCF clusters work with this approach, but it was a slightly different case.

making a node selector OS bound makes sense only as a temporary measure.
cc @kubernetes/sig-windows-misc

**Comment by timothysc** [comment]
@daschott This requires a KEP and affects multiple SIGs. 
There needs to be: 
- KEP outlining long term strategy 
- Buy-in from both sig-cluster-lifecycle and sig-node 
- Test automation!!!! 
- etc.

**Comment by michmike** [comment]
cc @PatrickLang  for our RuntimePolicy discussion

**Comment by neolit123** [comment]
going back to @chrisohaver 's [comment](https://github.com/kubernetes/kubernetes/pull/69940#issuecomment-438721206).

> As I understand it, there are some hurdles in generally getting K8s cluster DNS on Windows (not coredns specific). A significant one is that Windows systems don't have an /etc/resolv.conf, and kubelet is geared toward assuming all pods do. So there needs to be some way of informing windows Pods of the DNS policy.

it feels to me that we might want to get a DNS solution working on Windows eventually.
node selectors seem like a temporary solution.

**Comment by PatrickLang** [comment]
Dns works on Windows pods and is a function of CNI doing the right thing. 

CoreDns is used as a Kubernetes service. Windows pods can simply do dns lookups using that service IP and it will route to a Linux node running a matching pod for that service. Because CoreDns doesn't have a Windows container published, it needs the nodeselector.

**Comment by PatrickLang** [comment]
cc @yujuhong I discussed with SIG-Node today, and they don't have any concern with this as a point in time solution. NodeSelectors are widely used, and the labels needed in this PR are stable as of v1.14 #73048.

Long term, we could remove this node selector if one of the following is done

1. CoreDNS is published as a multi-arch, multi-OS image
2. Scheduler is updated to automatically pick nodes based on OS/arch (or other criteria) without NodeSelector. This could happen on its own, or if RuntimeClass is updated to factor in os/arch

**Review by neolit123 — COMMENTED** [review]
> I discussed with SIG-Node today, and they don't have any concern with this as a point in time solution. NodeSelectors are widely used, and the labels needed in this PR are stable as of v1.14 #73048.

> Long term, we could remove this node selector if one of the following is done

> CoreDNS is published as a multi-arch, multi-OS image
> Scheduler is updated to automatically pick nodes based on OS/arch (or other criteria) without NodeSelector. This could happen on its own, or if RuntimeClass is updated to factor in os/arch

SGTM, 
will leave the approval to @timothysc

**Review by timothysc — APPROVED** [review]
/approve 
/lgtm

**Comment by michmike** [comment]
@timothysc  could you please remove the hold so that it can be merged in?

**Pull request merged**
